duckdb-vortex/src/vortex_extension.cpp

```diff
@@ -204,16 +161,143 @@ static unique_ptr<FunctionData> VortexBind(ClientContext &context, TableFunction
+	}
+	for (auto idx = 0; idx < bind_data.columns_types.size(); idx++) {
+		auto col_name = bind_data.column_names[idx];
+		auto col_type = bind_data.columns_types[idx];
+		if (col_name != column_names[idx]) {
+			throw FatalException("Vortex file contains a column with a different name to the first");
+		}
+		if (col_type != column_types[idx]) {
+			throw FatalException("Vortex file contains a column with a different type to the first");
+		}
+	}
+}
+
+static File *OpenFileAndVerify(const std::string &filename, const VortexBindData &bind_data) {
+	auto new_column_names = vector<string>();
+	new_column_names.reserve(bind_data.column_names.size());
+	auto new_column_types = vector<LogicalType>();
+	new_column_names.reserve(bind_data.columns_types.size());
+
+	auto file = OpenFile(EnsureFileProtocol(filename), new_column_types, new_column_names);
+	VerifyNewFile(bind_data, new_column_types, new_column_names);
+	return file;
+}
+
+static ArrayStream *OpenArrayStream(const VortexBindData &bind_data, VortexScanGlobalState &global_state, File *file) {
+
+	auto options = FileScanOptions {
+	    .projection = global_state.projected_column_names.data(),
+	    .projection_len = static_cast<int>(global_state.projected_column_names.size()),
+	    .filter_expression = global_state.filter_str.data(),
+	    .filter_expression_len = static_cast<int>(global_state.filter_str.length()),
+	    // This is a multiple of the 2048 duckdb vector size, it needs tuning
+	    // This has a few factor effecting it:
+	    //  1. A smaller value means for work for the vortex file reader.
+	    //  2. A larger value reduces the parallelism available to the scanner
+	    .split_by_row_count = 2048 * 8,
+	};
+
+	return File_scan(file, &options);
+}
+
+static void VortexScanFunction(ClientContext &context, TableFunctionInput &data, DataChunk &output) {
+	auto &bind_data = data.bind_data->Cast<VortexBindData>();              // NOLINT
+	auto &global_state = data.global_state->Cast<VortexScanGlobalState>(); // NOLINT
+	auto &local_state = data.local_state->Cast<VortexScanLocalState>();    // NOLINT
+
+	if (local_state.array == nullptr) {
+		auto &slot = global_state.file_slots[local_state.thread_id];
+		std::lock_guard _l(slot.slot_lock);
+
+		if (global_state.finished.load()) {
+			output.SetCardinality(0);
+			return;
+		}
+
+		// 1. check we can make progress on current owned file
+		// 2. check we can get another file
+		// todo: 3. check if we can work steal from another thread
+		// 4. we are done
+
+		auto next = slot.array_stream != nullptr ? FFIArrayStream_next(slot.array_stream) : false;
+		if (!next) {
+			if (slot.array_stream != nullptr) {
+				FFIArrayStream_free(slot.array_stream);
+				slot.array_stream = nullptr;
+			}
+
+			auto file_idx = global_state.next_file.fetch_add(1);
+
+			if (file_idx >= global_state.expanded_files.size()) {
+				local_state.finished = true;
+				global_state.finished.exchange(true);
```

Comment:
Strictly speaking assignment is enough `= true` rather than `exchange` which returns the previous value. That returned value is unused here.